:sparkles: Output Pins

FluffyFTW opened the pull request:
- Added GPIO registers
- Added Output pins
> [!WARNING]
> NOT TESTED

FluffyFTW committed:
:sparkles: Initial drop of software

- Added power settings
- Added GPIO configs
- Added output pin constructor

FluffyFTW committed:
:sparkles: Initial drop of software

- Added power settings
- Added GPIO configs
- Added output pin constructor

FluffyFTW committed:
I want to make gpio_reg as a parameter for pin, how should I go about it

FluffyFTW committed:
I want to make gpio_reg as a parameter for pin, how should I go about it

FluffyFTW committed:
pin.hpp

FluffyFTW committed:
pin.hpp

FluffyFTW committed:
rebasing

FluffyFTW committed:
Merge branch 'main' of github.com:FluffyFTW/libhal-stm32f4

FluffyFTW committed:
:sparkles: Output pins

- Added GPIO regs
- Added Output pins
- [!WARNING] THIS HAS NOT BEEN TESTED

kammce commented on src/pin.cpp:
```diff
@@ -0,0 +1,72 @@
+  auto port_reg = get_reg(m_port);
+  auto pin_mode_mask =
+    hal::bit_mask::from<2 * m_pin, 2 * m_pin + 1>(port_reg->pin_mode);
+  uint8_t alt_func = 0;
```
```suggestion
    hal::bit_mask::from<2 * m_pin, 2 * m_pin + 1>(port_reg->pin_mode);
  uint8_t alt_func = 0;
```

This shouldn't work because m_pin is a runtime value and templates need compile time constants.

kammce commented on linker_scripts/libhal-stm32f4/stm32f411re.ld:
```diff
@@ -0,0 +1,7 @@
+__flash       = 0x08000000;
+__flash_size  = 512K;
+__ram         = 0x20020000;
+__ram_size    = 128K;
```
Is this correct?

kammce commented:
Also, the warning should be in the repo's README page. Otherwise no one will see it here.

FluffyFTW committed:
:sparkles: Implemented pin.hpp

- implemented pin.hpp
- Updated CMakeList.txt

FluffyFTW committed:
:sparkles: output pins

- implemented powering on periferals
- port address mapping now uses the letter and memory block location
- update readme to reflect the test conditions

FluffyFTW commented:
blinking video:
https://youtu.be/LWUTZw5LpPY?si=lfIrwRdwrz8H1tOL

kammce commented:
Fix CI and I'll take a look. Also, I cannot review this until its mergable. Please fix the merge conflicts.

kammce reviewed (changes requested):
Fix CI

FluffyFTW committed:
:white_check_mark: Fixed CI

kammce reviewed (approved):
We are going to merge this in as is an move it to 3.0.0
